File: ios/Sendro/Core/SendroClient.swift
//
//  SendroClient.swift
//  Sendro
//
//  Typed HTTP client for one Sendro host (http://ip:port), PROTOCOL.md v1.
//

import Foundation

enum SendroClientError: Error, LocalizedError {
    case badURL
    case badResponse
    case http(status: Int, code: String?, message: String?)
    case decoding

    var errorDescription: String? {
        switch self {
        case .badURL:
            return "Invalid host address."
        case .badResponse:
            return "Unexpected response from host."
        case .http(let status, let code, let message):
            // Never leak a bare status code to a person: HostStatus owns the
            // wording for every code this protocol can produce.
            return HostStatus.clientMessage(status: status, code: code, message: message)
        case .decoding:
            return "Could not read the host's response."
        }
    }

    var apiCode: String? {
        if case .http(_, let code, _) = self { return code }
        return nil
    }

    var httpStatus: Int? {
        if case .http(let status, _, _) = self { return status }
        return nil
    }

    /// The host's own `message` field, when it sent one. Distinguishes the
    /// two 503 flavours ("transfers paused" vs "transfer slots busy").
    var apiMessage: String? {
        if case .http(_, _, let message) = self { return message }
        return nil
    }
}

struct SendroClient {

    let baseURL: URL
    var token: String?

    /// Shared session for API calls (downloads use their own session).
    /// Fail-fast on purpose: ping / accept / status should error immediately
    /// when the host is unreachable so callers can react.
    private static let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.waitsForConnectivity = false
        config.timeoutIntervalForRequest = 40   // long poll 25s + slack
        config.timeoutIntervalForResource = 120
        config.httpAdditionalHeaders = ["Accept": "application/json"]
        return URLSession(configuration: config)
    }()

    /// Session dedicated to the outbox long poll. Unlike the fail-fast API
    /// session it waits for connectivity, so a transient Wi-Fi blip parks the
    /// poll instead of erroring it into backoff; the resource timeout still
    /// bounds every attempt so the poll loop can never hang forever.
    private static let pollSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.waitsForConnectivity = true
        config.timeoutIntervalForRequest = 40   // long poll 25s + slack
        config.timeoutIntervalForResource = 90  // caps time spent waiting for connectivity
        config.httpAdditionalHeaders = ["Accept": "application/json"]
        return URLSession(configuration: config)
    }()

    init?(host: String, port: UInt16, token: String? = nil) {
        // Be forgiving about what lands here (manual entry, resolver quirks):
        // trim whitespace, drop a pasted scheme/trailing slash, strip any
        // interface scope ("%en0"), and bracket bare IPv6 literals.
        var cleaned = host.trimmingCharacters(in: .whitespacesAndNewlines)
        for scheme in ["http://", "https://"] where cleaned.lowercased().hasPrefix(scheme) {
            cleaned = String(cleaned.dropFirst(scheme.count))
        }
        if let slash = cleaned.firstIndex(of: "/") { cleaned = String(cleaned[..<slash]) }
        if !cleaned.hasPrefix("["), let pct = cleaned.firstIndex(of: "%") {
            cleaned = String(cleaned[..<pct])
        }
        if cleaned.contains(":") && !cleaned.hasPrefix("[") {
            // Either a bare IPv6 literal, or "ip:port" pasted into the host
            // field. Two or more colons → IPv6, bracket it. Exactly one
            // colon → treat the tail as a (discarded) port.
            let colonCount = cleaned.filter { $0 == ":" }.count
            if colonCount >= 2 {
                cleaned = "[\(cleaned)]"
            } else if let colon = cleaned.firstIndex(of: ":") {
                cleaned = String(cleaned[..<colon])
            }
        }
        guard !cleaned.isEmpty, let url = URL(string: "http://\(cleaned):\(port)") else { return nil }
        self.baseURL = url
        self.token = token
    }

    // MARK: - Endpoints

    func info() async throws -> InfoResponse {
        try await get("/api/v1/info", timeout: 8)
    }

    func pairStart(_ request: PairStartRequest) async throws -> PairStartResponse {
        try await post("/api/v1/pair/start", body: request, timeout: 10)
    }

    func pairConfirm(_ request: PairConfirmRequest) async throws -> PairConfirmResponse {
        try await post("/api/v1/pair/confirm", body: request, timeout: 10)
    }

    func ping() async throws -> PingResponse {
        try await get("/api/v1/ping", timeout: 6)
    }

    func outboxLongPoll(waitSeconds: Int = 25) async throws -> OutboxResponse {
        try await get("/api/v1/outbox?waitSeconds=\(waitSeconds)",
                      timeout: TimeInterval(waitSeconds + 12),
                      session: SendroClient.pollSession)
    }

    @discardableResult
    func accept(transferId: String) async throws -> OkResponse {
        try await post("/api/v1/transfers/\(transferId)/accept", timeout: 10)
    }

    @discardableResult
    func reject(transferId: String) async throws -> OkResponse {
        try await post("/api/v1/transfers/\(transferId)/reject", timeout: 10)
    }

    @discardableResult
    func reportStatus(transferId: String, _ report: StatusReport) async throws -> OkResponse {
        try await post("/api/v1/transfers/\(transferId)/status", body: report, timeout: 10)
    }

    /// §11.2 — send an ephemeral text message to the host. Nothing about the
    /// text is cached or persisted here; it exists only for the duration of
    /// this request.
    @discardableResult
    func sendMessage(text: String) async throws -> OkResponse {
        do {
            let response: OkResponse = try await post("/api/v1/messages",
                                                      body: SendMessageRequest(text: text),
                                                      timeout: 15)
            return response
        } catch let error as SendroClientError {
            if error.httpStatus == 413 {
                throw SendroClientError.http(
                    status: 413,
                    code: "bad_request",
                    message: "Message too long — the limit is 32 KB of text.")
            }
            throw error
        }
    }

    /// Build the GET request for the file bytes (PROTOCOL.md §6.4).
    /// `rangeStart` != nil resumes from that offset with an If-Range guard.
    func makeFileRequest(transferId: String, rangeStart: Int64?, sha256: String) -> URLRequest {
        var request = URLRequest(url: baseURL.appendingPathComponent("api/v1/transfers/\(transferId)/file"))
        request.httpMethod = "GET"
        applyAuth(&request)
        if let rangeStart, rangeStart > 0 {
            request.setValue("bytes=\(rangeStart)-", forHTTPHeaderField: "Range")
            request.setValue("\"\(sha256.lowercased())\"", forHTTPHeaderField: "If-Range")
        }
        request.timeoutInterval = 60
        return request
    }

    /// Build the POST request for the §7 reverse upload (iPhone → PC).
    /// Raw body — NOT multipart; the caller attaches the file bytes via
    /// URLSession.uploadTask(with:fromFile:), which streams from disk.
    func makeUploadRequest(fileName: String, sha256Hex: String, contentLength: Int64) -> URLRequest {
        var request = URLRequest(url: baseURL.appendingPathComponent("api/v1/upload"))
        request.httpMethod = "POST"
        applyAuth(&request)
        request.setValue("UTF-8''" + Self.rfc5987Encode(fileName),
                         forHTTPHeaderField: "X-Sendro-File-Name")
        request.setValue(sha256Hex.lowercased(), forHTTPHeaderField: "X-Sendro-Sha256")
        // URLSession also derives Content-Length from the upload file; set it
        // explicitly so the header is present even before the body streams.
        request.setValue(String(contentLength), forHTTPHeaderField: "Content-Length")
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 60        // idle timeout between sent chunks
        return request
    }

    /// RFC 5987 value-chars encoding of a UTF-8 string. attr-chars stay
    /// literal — ALPHA / DIGIT and !#$&+-.^_`|~ — every other byte becomes
    /// %XX (uppercase hex) of its UTF-8 encoding. Note: this is deliberately
    /// NOT CharacterSet.alphanumerics (which admits non-ASCII letters).
    static func rfc5987Encode(_ value: String) -> String {
        var out = String()
        out.reserveCapacity(value.utf8.count)
        for byte in value.utf8 {
            if Self.rfc5987AttrChars.contains(byte) {
                out.append(Character(UnicodeScalar(byte)))
            } else {
                out += String(format: "%%%02X", Int(byte))
            }
        }
        return out
    }

    private static let rfc5987AttrChars: Set<UInt8> = {
        var set = Set<UInt8>()
        set.formUnion(UInt8(ascii: "a")...UInt8(ascii: "z"))
        set.formUnion(UInt8(ascii: "A")...UInt8(ascii: "Z"))
        set.formUnion(UInt8(ascii: "0")...UInt8(ascii: "9"))
        set.formUnion("!#$&+-.^_`|~".utf8)
        return set
    }()

    // MARK: - Plumbing

    private func applyAuth(_ request: inout URLRequest) {
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
    }

    private func get<Response: Decodable>(_ path: String,
                                          timeout: TimeInterval,
                                          session: URLSession = SendroClient.session) async throws -> Response {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw SendroClientError.badURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = timeout
        applyAuth(&request)
        return try await execute(request, session: session)
    }

    private func post<Response: Decodable>(_ path: String, timeout: TimeInterval) async throws -> Response {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw SendroClientError.badURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = timeout
        applyAuth(&request)
        return try await execute(request)
    }

    private func post<Body: Encodable, Response: Decodable>(
        _ path: String, body: Body, timeout: TimeInterval
    ) async throws -> Response {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw SendroClientError.badURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = timeout
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        applyAuth(&request)
        return try await execute(request)
    }

    private func execute<Response: Decodable>(_ request: URLRequest,
                                              session: URLSession = SendroClient.session) async throws -> Response {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw SendroClientError.badResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            if let apiError = try? JSONDecoder().decode(ApiError.self, from: data) {
                throw SendroClientError.http(status: http.statusCode,
                                             code: apiError.error,
                                             message: apiError.message)
            }
            throw SendroClientError.http(status: http.statusCode, code: nil, message: nil)
        }
        do {
            return try JSONDecoder().decode(Response.self, from: data)
        } catch {
            throw SendroClientError.decoding
        }
    }
}

File: android/app/src/main/java/com/sendro/android/core/SendroClient.kt
package com.sendro.android.core

import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Headers
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume

/**
 * Typed HTTP against ONE Sendro host (`http://ip:port`), PROTOCOL.md v1.
 *
 * WHY OKHTTP (and not Ktor): Ktor's client on Android still needs an engine,
 * and the only sane engine here is OkHttp — so Ktor would be OkHttp plus a
 * second abstraction. Everything Sendro needs is OkHttp's bread and butter:
 * per-call timeouts, real streaming bodies (`source()` / `RequestBody` from a
 * File), Range headers, connection pooling, and cancellation that actually
 * closes the socket.
 *
 * Timeout posture mirrors the iOS client's two URLSessions:
 *   * [Clients.api]  — fail fast. ping/accept/status must error immediately
 *     when the host is gone so the poll loop can react.
 *   * [Clients.poll] — a dedicated client with a long read timeout for the
 *     §6.2 long poll (waitSeconds + slack).
 *   * [Clients.transfer] — long-lived body streaming: a short connect timeout
 *     but a read timeout measured in "gap between chunks", not total time,
 *     and NO call timeout, so an 8 GB file is not cut off mid-flight.
 */
class SendroClient private constructor(
    val baseUrl: HttpUrl,
    private val token: String?,
) {

    companion object {

        /**
         * Be forgiving about what lands here (manual entry, resolver quirks):
         * trim, drop a pasted scheme/path, strip an interface scope ("%wlan0"),
         * bracket bare IPv6 literals, and tolerate "ip:port" pasted into the
         * host field.
         *
         * @return null when nothing usable can be made of [host].
         */
        fun create(host: String, port: Int, token: String? = null): SendroClient? {
            var cleaned = host.trim()
            for (scheme in listOf("http://", "https://")) {
                if (cleaned.startsWith(scheme, ignoreCase = true)) {
                    cleaned = cleaned.substring(scheme.length)
                }
            }
            cleaned.indexOf('/').let { if (it >= 0) cleaned = cleaned.substring(0, it) }
            if (!cleaned.startsWith("[")) {
                cleaned.indexOf('%').let { if (it >= 0) cleaned = cleaned.substring(0, it) }
            }
            if (cleaned.contains(':') && !cleaned.startsWith("[")) {
                val colons = cleaned.count { it == ':' }
                cleaned = if (colons >= 2) {
                    "[$cleaned]"                       // bare IPv6 literal
                } else {
                    cleaned.substringBefore(':')       // "ip:port" pasted in
                }
            }
            if (cleaned.isEmpty() || port !in 1..65535) return null
            val url = "http://$cleaned:$port/".toHttpUrlOrNull() ?: return null
            return SendroClient(url, token)
        }

        private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()
        private val OCTET_MEDIA = "application/octet-stream".toMediaType()

        /**
         * RFC 5987 value-chars encoding of a UTF-8 string. attr-chars stay
         * literal — ALPHA / DIGIT and `!#$&+-.^_\`|~` — every other byte
         * becomes %XX (uppercase hex) of its UTF-8 encoding.
         *
         * Deliberately NOT "isLetterOrDigit", which admits non-ASCII letters
         * and would leave `ç` unencoded.
         */
        fun rfc5987Encode(value: String): String {
            val out = StringBuilder(value.length + 16)
            for (byte in value.toByteArray(Charsets.UTF_8)) {
                val v = byte.toInt() and 0xFF
                val c = v.toChar()
                val literal = (c in 'a'..'z') || (c in 'A'..'Z') || (c in '0'..'9') ||
                    c in "!#\$&+-.^_`|~"
                if (literal) {
                    out.append(c)
                } else {
                    out.append('%')
                    out.append("0123456789ABCDEF"[v ushr 4])
                    out.append("0123456789ABCDEF"[v and 0x0F])
                }
            }
            return out.toString()
        }
    }

    /**
     * Process-wide OkHttp clients. One dispatcher and one connection pool for
     * the whole app; the three variants differ only in timeouts.
     */
    object Clients {
        private val base: OkHttpClient = OkHttpClient.Builder()
            .retryOnConnectionFailure(true)
            .connectTimeout(6, TimeUnit.SECONDS)
            .build()

        /** Fail-fast API calls: info, pair, ping, accept, reject, status. */
        val api: OkHttpClient = base.newBuilder()
            .readTimeout(12, TimeUnit.SECONDS)
            .writeTimeout(12, TimeUnit.SECONDS)
            .callTimeout(20, TimeUnit.SECONDS)
            .build()

        /**
         * §6.2 long poll. `waitSeconds` is capped at 30 by the host, so a
         * 55 s read timeout leaves generous slack while still bounding a poll
         * that will never answer.
         */
        val poll: OkHttpClient = base.newBuilder()
            .readTimeout(55, TimeUnit.SECONDS)
            .callTimeout(70, TimeUnit.SECONDS)
            .build()

        /**
         * File bytes in either direction. `readTimeout` is the maximum gap
         * between chunks, not the total transfer time, and `callTimeout` is
         * explicitly 0 (unbounded) so a multi-gigabyte file is never cut off.
         */
        val transfer: OkHttpClient = base.newBuilder()
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
            .callTimeout(0, TimeUnit.MILLISECONDS)
            .build()
    }

    // -----------------------------------------------------------------------
    // Endpoints
    // -----------------------------------------------------------------------

    /** §5 */
    suspend fun info(): InfoResponse = get("api/v1/info")

    /** §4.1 */
    suspend fun pairStart(request: PairStartRequest): PairStartResponse =
        post("api/v1/pair/start", SendroJson.encodeToString(request))

    /** §4.2 */
    suspend fun pairConfirm(request: PairConfirmRequest): PairConfirmResponse =
        post("api/v1/pair/confirm", SendroJson.encodeToString(request))

    /** §4.3 */
    suspend fun ping(): PingResponse = get("api/v1/ping")

    /** §6.2 — long poll on the dedicated client. */
    suspend fun outboxLongPoll(waitSeconds: Int = 25): OutboxResponse =
        get("api/v1/outbox?waitSeconds=$waitSeconds", Clients.poll)

    /** §6.3 */
    suspend fun accept(transferId: String): OkResponse =
        post("api/v1/transfers/$transferId/accept", null)

    /** §6.3 */
    suspend fun reject(transferId: String): OkResponse =
        post("api/v1/transfers/$transferId/reject", null)

    /** §6.5 */
    suspend fun reportStatus(transferId: String, report: StatusReport): OkResponse =
        post("api/v1/transfers/$transferId/status", SendroJson.encodeToString(report))

    /**
     * §11.2 — send one ephemeral text message. Nothing about the text is
     * cached or persisted here; it exists only for the duration of this call.
     */
    suspend fun sendMessage(text: String): OkResponse = try {
        post("api/v1/messages", SendroJson.encodeToString(SendMessageRequest(text)))
    } catch (e: SendroHttpException) {
        if (e.status == 413) {
            throw SendroHttpException(
                413, "bad_request",
                "Message too long — the limit is 32 KB of text.",
            )
        }
        throw e
    }

    /**
     * §6.4 — the GET for the actual bytes. [rangeStart] > 0 resumes from that
     * offset with an `If-Range` guard against the file having changed.
     *
     * Returns the request; the caller executes it on [Clients.transfer] and
     * streams the body itself.
     */
    fun fileRequest(transferId: String, rangeStart: Long, sha256: String): Request {
        val builder = Request.Builder()
            .url(baseUrl.newBuilder().addPathSegments("api/v1/transfers/$transferId/file").build())
            .get()
            .header("Accept", "*/*")
            // Belt and braces: the host already sends identity, and asking for
            // it explicitly means no proxy on the path can helpfully gzip the
            // bytes we are about to hash.
            .header("Accept-Encoding", "identity")
        if (rangeStart > 0) {
            builder.header("Range", "bytes=$rangeStart-")
            builder.header("If-Range", "\"${sha256.lowercase()}\"")
        }
        applyAuth(builder)
        return builder.build()
    }

    /**
     * §7 — the reverse upload (phone -> PC). Raw body, NOT multipart.
     *
     * The body streams from [file] in 1 MiB chunks and reports the running
     * total through [onBytesSent], which is the only place OkHttp exposes
     * upload progress at all. `Content-Length` comes from `File.length()`, so
     * the header is correct before a single byte moves and memory stays flat
     * for an 8 GB video exactly as it does for an 8 KB note.
     */
    fun uploadRequest(
        file: File,
        fileName: String,
        sha256Hex: String,
        onBytesSent: (Long) -> Unit = {},
    ): Request {
        val builder = Request.Builder()
            .url(baseUrl.newBuilder().addPathSegments("api/v1/upload").build())
            .header("X-Sendro-File-Name", "UTF-8''" + rfc5987Encode(fileName))
            .header("X-Sendro-Sha256", sha256Hex.lowercase())
        applyAuth(builder)
        builder.post(StreamingFileBody(file, OCTET_MEDIA, onBytesSent))
        return builder.build()
    }

    // -----------------------------------------------------------------------
    // Plumbing
    // -----------------------------------------------------------------------

    private fun applyAuth(builder: Request.Builder) {
        token?.let { builder.header("Authorization", "Bearer $it") }
    }

    private suspend inline fun <reified T> get(
        path: String,
        client: OkHttpClient = Clients.api,
    ): T {
        val url = baseUrl.resolve(path) ?: throw SendroHttpException(0, "bad_request", "Bad URL")
        val builder = Request.Builder().url(url).get().header("Accept", "application/json")
        applyAuth(builder)
        return decode(execute(builder.build(), client))
    }

    private suspend inline fun <reified T> post(
        path: String,
        jsonBody: String?,
        client: OkHttpClient = Clients.api,
    ): T {
        val url = baseUrl.resolve(path) ?: throw SendroHttpException(0, "bad_request", "Bad URL")
        val body: RequestBody = jsonBody?.toRequestBody(JSON_MEDIA)
            ?: ByteArray(0).toRequestBody(null)
        val builder = Request.Builder().url(url).post(body).header("Accept", "application/json")
        applyAuth(builder)
        return decode(execute(builder.build(), client))
    }

    /** Reads the whole (small, JSON) body; never used for file bytes. */
    private suspend fun execute(request: Request, client: OkHttpClient): String {
        val response = client.newCall(request).awaitResponse()
        return response.use { r ->
            val text = r.body?.string().orEmpty()
            if (!r.isSuccessful) throw errorFor(r.code, text, r.header("Retry-After"))
            text
        }
    }
}

/**
 * A raw file body that streams from disk and counts as it goes.
 *
 * OkHttp's own `File.asRequestBody()` would stream just as well but gives no
 * progress; this is the same loop with a callback. `writeTo` can be invoked
 * more than once if OkHttp retries the request, so the counter restarts each
 * time and the callback always reports an absolute total.
 */
private class StreamingFileBody(
    private val file: File,
    private val mediaType: okhttp3.MediaType,
    private val onBytesSent: (Long) -> Unit,
) : RequestBody() {

    override fun contentType(): okhttp3.MediaType = mediaType

    override fun contentLength(): Long = file.length()

    override fun writeTo(sink: okio.BufferedSink) {
        var total = 0L
        onBytesSent(0L)
        file.inputStream().use { input ->
            val buffer = ByteArray(SendroCrypto.CHUNK_BYTES)
            while (true) {
                val read = input.read(buffer)
                if (read <= 0) break
                sink.write(buffer, 0, read)
                total += read
                onBytesSent(total)
            }
        }
        sink.flush()
    }
}

/** JSON decode with a uniform failure, so callers only handle one error type. */
private inline fun <reified T> decode(text: String): T =
    try {
        SendroJson.decodeFromString<T>(text)
    } catch (_: Exception) {
        throw SendroHttpException(0, "bad_response", "Could not read the host's response.")
    }

/**
 * Suspend bridge for an OkHttp call. Cancelling the coroutine cancels the call
 * (which closes the socket) — the property the poll loop and the transfer
 * cancel button both depend on.
 */
suspend fun Call.awaitResponse(): Response = suspendCancellableCoroutine { continuation ->
    continuation.invokeOnCancellation { runCatching { cancel() } }
    enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            if (continuation.isActive) continuation.resumeWith(Result.failure(e))
        }

        override fun onResponse(call: Call, response: Response) {
            // If the coroutine is already gone nobody will close this body —
            // do it here rather than leaking the connection.
            if (continuation.isActive) continuation.resume(response) else response.closeQuietly()
        }
    })
}

internal fun Response.closeQuietly() {
    runCatching { close() }
}

/**
 * §9 — an HTTP error carrying everything the state machine needs.
 *
 * [retryAfterSeconds] is the host's own `Retry-After`, already clamped. It is
 * the difference between "this transfer failed" and "the host told us when to
 * come back", and the whole 503 story depends on it surviving up to the
 * engine rather than being flattened into a string here.
 */
class SendroHttpException(
    val status: Int,
    val code: String?,
    val serverMessage: String?,
    /** Clamped `Retry-After`, or null when the host did not send a usable one. */
    val retryAfterSeconds: Int? = null,
) : IOException(
    // Even the exception's own message must be readable: it can reach the UI
    // through `sendroMessage()` on paths that have no peer name to hand.
    when {
        !serverMessage.isNullOrBlank() -> serverMessage
        status > 0 -> HttpSemantics.explain(status, serverMessage, "The other device", receiving = true)
        else -> "The request was refused."
    },
) {
    val disposition: HttpDisposition get() = HttpSemantics.disposition(status)

    val busyReason: BusyReason get() = HttpSemantics.busyReason(serverMessage)

    /** Plain language for the UI. Never contains a status number. */
    fun explain(peerName: String, receiving: Boolean): String =
        HttpSemantics.explain(status, serverMessage, peerName, receiving)
}

/**
 * Builds the typed exception for a non-2xx response, keeping the host's
 * `Retry-After` intact.
 */
internal fun errorFor(status: Int, body: String, retryAfter: String?): SendroHttpException {
    val parsed = runCatching { SendroJson.decodeFromString<ApiError>(body) }.getOrNull()
    return SendroHttpException(
        status = status,
        code = parsed?.error,
        serverMessage = parsed?.message,
        retryAfterSeconds = HttpSemantics.retryAfterSeconds(retryAfter),
    )
}

/** Human text for anything thrown by the client, for direct UI display. */
fun Throwable.sendroMessage(): String =
    message?.takeIf { it.isNotBlank() } ?: (this::class.simpleName ?: "Unknown error")

/** Headers helper: the §6.4 `Content-Range` total, when the host sent one. */
fun Headers.contentRangeTotal(): Long? {
    val value = this["Content-Range"] ?: return null
    val slash = value.lastIndexOf('/')
    if (slash < 0) return null
    return value.substring(slash + 1).trim().toLongOrNull()
}
